File: Ruby/프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.rb
def isBreadNow(ingre)
    if ingre == 1
        return 1
    end
    return 0
end




def solution(ingredient)
    answer = 0
    

    # cntStack 을 통해서 처리    
    
    stack = []
    cntStack = []
    reciepe = [1,2,3,1]
    cnt=0

    nowCheck = false

    ingredient.each_with_index{
        |ingre, index|
    
        if ingre == reciepe[cnt]
            cnt+=1
            cntStack << cnt
            if cnt == 4
                answer+=1
                4.times{
                    cntStack.pop()
                }
                if cntStack.last.nil?
                    cnt =0
                else
                    cnt = cntStack.last
                end
            end
        else
            cnt = isBreadNow(ingre)
            cntStack << cnt
        end

    }
    
    return answer
end

File: Java/프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] ingredient) {
        int answer = 0;
        int[] stack = new int[ingredient.length];
        int index = 0;
        
        for (int ing : ingredient) {
            // 현재 재료를 스택에 추가
            stack[index++] = ing;
            
            // 스택에 4개 이상의 재료가 있는 경우 확인
            if (index >= 4) {
                // 마지막 4개가 햄버거 패턴인지 확인
                if (stack[index - 4] == 1 && 
                    stack[index - 3] == 2 && 
                    stack[index - 2] == 3 && 
                    stack[index - 1] == 1) {
                    
                    // 햄버거가 완성되면 포인터를 4 감소시킴 (4개 요소 제거 효과)
                    index -= 4;
                    answer++;
                }
            }
        }
        
        return answer;
    }
}